### PR TITLE
Include non-competing staff in the public WCIF export

### PR DESCRIPTION
**Pull request by timreyn**
This will allow tools like competitiongroups.com to show non-competing staff members' assignments.

Depends on thewca/wcif#15.

### CODE REVIEW

**Review by gregorbg — CHANGES_REQUESTED** [review]
Formally blocking this implementation until the design discussion in the accompanying WCIF PR has reached a conclusion

**Comment by timreyn** [comment]
Per the discussion on thewca/wcif#15, I switched this to just use "accepted" rather than a new status.

I did not add a "roles: ['staff']", since it feels wrong to include that only for non-competing staff. I think that if this is desired, it should be through a more systematic marking of which competitors are staff.

Another downside is that this includes non-human accounts like CubingUSA Treasurer. As far as I know there is nothing in the database to indicate which accounts are humans, but I filed #8011 to track this.

**Review by gregorbg — COMMENTED** [review]
Why are non-competing staff registrations not marked as `accepted?` anyways? Shouldn't they be accepted in the competitor waitlist anyways, making this a zero line code change?

**Comment by timreyn** [comment]
@gregorbg : There's a few reasons I'd prefer not tracking `accepted` for staff currently:

1. If in the future we have a formal UI for registering to be staff, then I think it would make sense to have a parallel accepted / waitlist / denied structure for staff, which would include the possibility of someone being accepted to compete but denied to staff. There are lots of considerations involved in that, such as whether that should be a separate `staff` table, versus a single entry for both staff and volunteers, or if not then making sure all of the possible state transitions are handled reasonably.
2. In the meantime, marking staff as `accepted?` (the same field as competitors) seems like it's likely that many different checks would need to be updated. e.g. in `competition.rb`, `registration_full?` and `registration_full_message` both reference the competitor limit. I'm worried that this would need to touch a lot of things, and that we'd miss something.
3. I think that, without a formal UI for managing staff on the WCA site, tracking staff as being accepted / waitlisted / denied on the WCA site (and doing all the work involved in 2) is overkill. Any competition that wants to push non-competing staff to the WCA site will probably be tracking the staff list (and accepted / waitlisted / denied) separately e.g. in Google Sheets.

Thoughts?

(holding off on adding isCompeting until the WCIF change is approved + merged)

**Comment by gregorbg** [comment]
Point number 2 is especially convincing. It is worth noting though that we are building the new registration microservice and UI with exactly the type of considerations from (1) in mind. I will approve the WCIF schema change very soon, feel free to start working on the implementation.

**Comment by timreyn** [comment]
Great! Looks like you merged the WCIF PR. I added the isCompeting field as discussed over there, and I also updated `update_persons_wcif` to only add registrations that have this field set.

Good to hear this is the kind of thing that the new registration flow is hoping to support! I also think the schema questions in (1) are worth thinking through as part of that process.

**Comment by timreyn** [comment]
(the failures look unrelated to my change)

**Comment by gregorbg** [comment]
> (the failures look unrelated to my change)

Yup, sorry! Restarted the CI, this is an old choke that we hope to fix once we're off `structure.sql` (which is possible now that PHP is dead, we just need to find time to actually do the change)

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:1775]: Rails has a nifty shorthand to make this more readable: `registration.present?` works even on `nil` values and essentially returns whether a value is nil or not (plus some other very arcane ActiveRecord edge cases which are not relevant here anyways)

Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:1775]: First part of the OR is obvious.

Third part should use `wcif_person["registration"]["isCompeting"]` in camel case instead of `is_competing` snake case. Also, at this point in time the WCIF has already been schema-checked (this is actually the first thing that happens when hitting the PATCH endpoint) so I believe it is fair to assume that `isCompeting` is always a boolean anyways. So does explicitly checking `!= false` make sense?!

In a wider sense, what about reworking this into `next if !registration.present? && wcif_person["registration"] && !wcif_person["registration"]["isCompeting"]`? I find it much easier to understand than the implicitly negated form baked into the `unless`, although that might just be personal preference. Totally open to debate and to being convinced of the current `unless` version.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:1645]: Might be worth adding a comment here why this extra `is_competing` check is necessary, for people who don't know that we (currently) don't mark non-competing staff as "accepted" in the UI.

File: WcaOnRails/app/models/registration.rb
Comment (gregorbg) [inline WcaOnRails/app/models/registration.rb:207]: Might be worth a comment, see above. On second thought, might even be worth a refactor into its own method in `Registration`, along the lines of `attending?`

**Review by gregorbg — CHANGES_REQUESTED** [review]
The only definitive change that we need is to use `isCompeting` camel case when reading from the WCIF. The rest of my remarks are opinions about code style which are open to debate.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition.rb:1775]: Hold on, now I confused myself with that suggestion about reworking the `next` condition. What are we even trying to achieve here? Sorry :see_no_evil: :woozy_face:

Comment (timreyn) [reply WcaOnRails/app/models/competition.rb:1775]: Oops, fixing to isCompeting.

What are we trying to achieve? The idea was to go back to the old behavior (see base side of https://github.com/thewca/worldcubeassociation.org/pull/7601/files) unless the app PATCHing explicitly says isCompeting: false. This means all the implicitly-added organizer entities won't show up as non-competing registrations anymore.

I was trying to explicitly handle a missing isCompeting differently than a set-to-false isCompeting (does the schema validation fail if isCompeting is missing?) My assumption was that missing isCompeting meant the app writer didn't consider that field, and set-to-false means they considered it and said "yes this person is not competing", but both nil and false would be treated as false. Thoughts?

Agree with reversing the negatives.

**Review by timreyn — COMMENTED** [review]
(responding to one comment now before I make changes + test)

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition.rb:1775]: The idea makes sense. So what about a two-step approach?
```ruby
# ...from before, nothing new:
registration = registrations.find { |reg| reg.user_id == wcif_person["wcaUserId"] }

adding_non_competing = wcif_person["registration"].present? && wcif_person["registration"]["is_competing"] == false
registration ||= ...whatever... if adding_non_competing

# similar to the old status quo from the PR you referenced, but with a tiny bit more readable `.present?` now
next unless registration.present?
```

File: WcaOnRails/app/models/competition.rb
Comment (timreyn) [reply WcaOnRails/app/models/competition.rb:1648]: Done.

File: WcaOnRails/app/models/competition.rb
Comment (timreyn) [reply WcaOnRails/app/models/competition.rb]: OK, done! Rubocop doesn't like `registration ||= ... if adding_non_competing`, so I went with a normal if statement to appease it.

Comment (timreyn) [reply WcaOnRails/app/models/competition.rb]: Done.

File: WcaOnRails/app/models/registration.rb
Comment (timreyn) [reply WcaOnRails/app/models/registration.rb]: Moved into its own method.

**Review by timreyn — COMMENTED** [review]
Thanks Gregor!

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition.rb:1648]: Now you have an extra method below but you don't use it up here...
What about `.select { |r, registrant_id| authorized || r.wcif_status == "accepted" }`?

Comment (timreyn) [reply WcaOnRails/app/models/competition.rb]: Sure, done.

**Review by gregorbg — APPROVED** [review]

**Pull request merged**
